### PR TITLE
Zero V3 - TypeScript, Babel, Jest!

### PR DESCRIPTION
**Pull request by ItsJonQ**
## Zero V3 - TypeScript, Babel, Jest!

This update brings out of the box **TypeScript** support into Zero.

### New Features

* TypeScript support
* Babel + TypeScript
* Jest + TypeScript
* Improved CLI
* Set up commands

```
📦  Zero

zero <command>

Example:
  zero build

Options:
  -V, --version     output the version number
  -h, --help        output usage information

Commands:
  build [options]   Builds project with Babel, Rollup, or TypeScript
  bundle [options]  Bundles project into single files with Rollup
  contributors      Generates markdown file with all contributors
  format [options]  Formats files with Prettier
  link [options]    Lints files with ESLint
  new               Generate a new module
  pre-commit        Lints files before staging for commit
  prestart          Automatically install dependencies before starting
  release           Publish to npm
  setup [options]   Sets up tooling in project
  test [options]    Run test with Jest
  typecheck         Check types with TypeScript
  validate          Validates project with lint, tests, and build
```

### Strategy

The Babel/Jest updates combine the simpliest/best config settings from [kcd-scripts](https://github.com/kentcdodds/kcd-scripts) and [create-react-app](https://github.com/facebook/create-react-app).

This set up allows you can compile your project with either Babel **or** TypeScript.

For additional details, check out the screencasts below:

### Feature Overview Video (3:10)

https://soapbox.wistia.com/videos/ccx8Wl6Qeg

### How TypeScript, Babel, and Jest work together Video (7:05)

https://soapbox.wistia.com/videos/FTQYVkePeR

### Testing

The easiest way to test this.. is to clone this project to your computer:
https://github.com/helpscout/wedux

Check out the `zero-3` branch:

And run some commands!

### CODE REVIEW

File: src/config/jest.config.js [hunk 1]
```diff
@@ -12,28 +18,52 @@ const ignores = [
...
+// kcd-script's setup was used as the foundation because of it's simplicity.
+// CRA's settings have been added for TypeScript support.
 const jestConfig = {
   roots: [fromRoot('src')],
   testEnvironment: ifAnyDep(['webpack', 'rollup', 'react'], 'jsdom', 'node'),
   testURL: 'http://localhost',
   moduleFileExtensions: ['js', 'jsx', 'json', 'ts', 'tsx'],
-  collectCoverageFrom: ['src/**/*.+(js|jsx|ts|tsx)'],
-  testMatch: ['**/__tests__/**/*.+(js|jsx|ts|tsx)'],
+  // From create-react-app
```
Comment (tinkertrain) [inline src/config/jest.config.js]: Pretty cool that you left these comments 👍

File: src/scripts/setup/babel.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    console.log('Cancelling Babel setup')
+    process.exit(0)
+  }
+
+  const content = '{ "presets": ["@helpscout/zero/babel"] }'
+
+  console.log(`Generating ${babelConfig}...`)
+  writeFileToRoot('./.babelrc', content)
+
+  console.log()
```
Comment (tinkertrain) [inline src/scripts/setup/babel.js]: Is this empty `console.log` intentional?

Comment (tinkertrain) [reply src/scripts/setup/babel.js]: I see a few, so maybe they are, curious about the intention!

Comment (ItsJonQ) [reply src/scripts/setup/babel.js]: @tinkertrain That's just to leave a new line `\n` in the log :)

So it would be

```
one

two
```

vs

```
one
two
```

**Comment by ItsJonQ** [comment]
@tinkertrain Thank you so much for the review!!!! 🙏

**Comment by ItsJonQ** [comment]
@knicklabs is out today (feel better soon! ❤️ ). Will merge this up

**Pull request merged**